README.md
# Tuya Embedded Gateway Testing SDK

[English version](README.md) | [中文版](README_cn.md)

## Introduction

### Directory

```shell
.
├── CHANGE_LOG.md  
├── Makefile
├── README.md
├── demo
│   └── demo.c
└── sdk
    ├── include
    │   ├── tuya_prod_test.h
    │   └── tuya_testframe_handle.h
    └── lib
        ├── shared
        │   ├── libtuya_prodtest.so
        │   └── libtuya_prodtest.so.stripped
        └── static
            ├── libtuya_prodtest.a
            └── libtuya_prodtest.a.stripped
```

### Library

This SDK provides the dynamic library, static library, strip library. And the static library is selected by default.

## Getting started

* Command for local compilation:

    ```
    make
    ```
* Command for local clear:

    ```
    make clean
    ```
* The path of the executable file:

    `output/bin`

## Demo reference

Please see the `demo` directory for details.

## Technical support

- [Tuya Developer](https://developer.tuya.com/en/)
- [Tuya Smart Documentation](https://developer.tuya.com/en/docs/iot)
- [Technical ticket](https://service.console.tuya.com)


CHANGELOG.md
# Tuya Embedded Gateway Product Test SDK Release Notes

[English](CHANGELOG.md) | [中文版](CHANGELOG_cn.md)

## v1.0.0

### Release Date

2020.4.26

### Highlights

- The first version of the production test SDK

<div>
        &copy; 2014 - 2021 Tuya Inc. All Rights Reserved.
</div>

Library/gcc/sdk/include/tuya_prod_test.h
#ifndef _TUYA_PROD_TEST_H_
#define _TUYA_PROD_TEST_H_


#ifdef __cplusplus
extern "C"
{
#endif
#include <stdio.h>
#include <stdbool.h>

//log
#define MAX_LOG_LEN 1024

int log_print(const char *flag_str, char *file, const char *func, unsigned int line, char *fmt, ...);


//#define TYPD_DEBUG(fmt, ...)  printf("PT DEBUG:[%s:%s:%d] "fmt"\r\n", __FILE__,__FUNCTION__,__LINE__,##__VA_ARGS__)
//#define TYPD_ERR(fmt, ...)  printf("PT ERR:[%s:%s:%d] "fmt"\r\n", __FILE__,__FUNCTION__,__LINE__,##__VA_ARGS__)
#define TYPD_DEBUG(fmt, ...)  log_print("PT DEBUG", __FILE__,__FUNCTION__,__LINE__,fmt,##__VA_ARGS__)
#define TYPD_ERR(fmt, ...)  log_print("PT ERR", __FILE__,__FUNCTION__,__LINE__,fmt,##__VA_ARGS__)


#define TYPD_LOG_PUT(a) printf("%s\r\n",a);

/**************************----define----*************************************/
//协议版本号
#define TUYA_PROD_PROTOCOL "1.0.0"
//sdk版本号
#define TUYA_PROD_SDK_VERSION "1.0.0"


//UDP PORT
#define TUYA_TEST_RECV_PORT 12130

//tuya产测帧长度 根据具体需求自定义
#define TUYA_FRAME_MIN_LEN 7
#define TUYA_FRAME_MAX_LEN 1048

#define IPADDR_TUYATEST_RCVE       ((unsigned int)0x00000000UL)

#define falseFrameResultStr "{\"ret\":false}"
#define trueFrameResultStr  "{\"ret\":true}"

//支持的按键数量
#define TYPT_MAX_KEY 100

/**************************----tcpser----*************************************/
//限制连接上位机为24
#define TYT_CLIENT_LMT 24 
typedef enum {
    TTSS_INIT = 0,
    TTSS_CR_SERV,
    TTSS_SELECT,
}TUYATEST_TCP_SERV_STAT_E;

typedef struct tuyatest_recv_buffer {
    unsigned char *r_buffer[TYT_CLIENT_LMT];
    int r_soc[TYT_CLIENT_LMT];          //接受buffer对应的套接字
    unsigned int handle_cnt;        //当前需要处理套接字的个数
    unsigned int current_handle_index; //当前处理的套接字对应数组中的索引
    unsigned int sum;               //等同于CLIENT_LMT
    int has_malloc;              //表示已经为r_buffer分配了内存
}TUYATEST_R_BUFFER;

//套接字管理结构
typedef struct client_soc_manage
{
    int ser_fd;
    int accep_soc_num;    
    int accep_soc[TYT_CLIENT_LMT];
    int has_recv_buffer[TYT_CLIENT_LMT];
}CLI_SOC_MANAGE;

/**************************----frame----*************************************/
typedef struct tuya_test_frame{
    unsigned char head[2];
    unsigned char version[1];
    unsigned char command[1];
    unsigned char data_len[2];
    unsigned char *data;
    unsigned char check_sum[1];
    unsigned int data_len_uint; //产测帧数据部分长度
    unsigned char *frame_buffer;//指向的buffer含有帧的所有数据，最后要释放掉
}TuyaTestFrame;

typedef enum
{
    TUYATEST_MODE = 0x00,
    TUYATEST_R_MASTER_FIRMWARE = 0x01,
    TUYATEST_R_SLAVE_FIRMWARE = 0x02,
    TUYATEST_W_CFGINFO = 0x03,
    TUYATEST_R_CFGINFO = 0x04,
    TUYATEST_W_MASTER_MAC = 0x05,
    TUYATEST_R_MASTER_MAC = 0x06,
    TUYATEST_W_SLAVE_MAC = 0x07,
    TUYATEST_R_SLAVE_MAC = 0x08,
    TUYATEST_SELF_TEST = 0x09,
    TUYATEST_BUTTON_TEST = 0x0a,
    TUYATEST_LED_TEST = 0x0b,    
    TUYATEST_W_BSN = 0x0c,
    TUYATEST_R_BSN = 0x0d,
    TUYATEST_W_SN = 0x0e,
    TUYATEST_R_SN = 0x0f,
    TUYATEST_MASTER_CAP = 0x10,
    TUYATEST_SLAVE_CAP = 0x11,    
    TUYATEST_SPEAKER = 0x15,      
    TUYATEST_W_COUNTRY = 0x1C, 
    TUYATEST_R_COUNTRY = 0x1D,
    TUYATEST_UPGRADE_CFG = 0x1e,
    TUYATEST_UPGRADE_START = 0x1f,
    TUYATEST_UPGRADEING = 0x20,
    TUYATEST_UPGRADE_END = 0x21,
    TUYATEST_GW_REBOOT = 0x22,    
    TUYATEST_SLAVE_UPGRADE = 0x23,        
    TUYATEST_LTE_TEST = 0x29,
    TUYATEST_FLAG_DECTECT = 0x2A,
    TUYATEST_LTE_READ_CCID = 0x2B,
} TUYATEST_CMD;

//传输数据的方式 目前支持tcp，可扩展
typedef enum
{
    TPM_TCP = 0,
    TPM_SERIAL,
    TPM_MAX
} TUYA_PRODTEST_MODE;

//产测线程运行状态
typedef enum
{
    TEST_THD_NORMAL = 1,
    TEST_THD_EXIT
} TUYA_TEST_THREAD_STATE;

/**************************----callback----*************************************/
typedef TuyaTestFrame * (*TUYA_GET_FRAME)(int buffer_index);
typedef void (*TUYA_PUT_FRAME)(unsigned char cmd,unsigned char *data);
typedef void (*TUYA_FREE_FRAME)(TuyaTestFrame *frame);
typedef void *(*thread_process)(void *pArg);

typedef struct tuya_frame_handler{
    TUYA_GET_FRAME tuya_get_frame;
    TUYA_PUT_FRAME tuya_put_frame;
    TUYA_FREE_FRAME tuya_free_frame;    
 }TF_HANDLER;

typedef int (*TYTEST_COMMON_HANDL)(TuyaTestFrame *frame, char *out_buf);
typedef int (*TYTEST_COMMON_ASYNC_HANDL)(int index,TuyaTestFrame *frame, char *out_buf);
typedef int (*TYTEST_COMMON_HANDL_NOBUF)(TuyaTestFrame *frame);
typedef int (*TYTEST_KEY_HANDL)(TuyaTestFrame *frame ,unsigned int *response_key_num);
typedef int (*TYTEST_KEY_ASYNC_HANDL)(TuyaTestFrame *frame ,int index);
typedef void (*TYTEST_CUSTOMIZE_HANDL)(TuyaTestFrame *frame, char *out_buf);
typedef void (*TYTEST_REBOOT_HANDL)(TuyaTestFrame *frame);

typedef struct {    
    TYTEST_COMMON_HANDL enter_prodtest_cb;              //1.enter prodtest
    TYTEST_COMMON_HANDL master_firm_cb;                 //2.master_firm
    TYTEST_COMMON_HANDL slave_firm_cb;                  //3.slave firm
    TYTEST_COMMON_HANDL_NOBUF w_cfg_cb;                 //4.write cfd info
    TYTEST_COMMON_HANDL r_cfg_cb;                       //5.read cfg info
    TYTEST_COMMON_HANDL_NOBUF w_master_mac_cb;          //6.write master mac
    TYTEST_COMMON_HANDL r_master_mac_cb;                //7.read master mac
    TYTEST_COMMON_HANDL_NOBUF w_slave_mac_cb;           //8.write slave mac
    TYTEST_COMMON_HANDL r_slave_mac_cb;                 //9.read slave mac
    TYTEST_COMMON_HANDL_NOBUF w_bsn_cb;                 //10.write bsn
    TYTEST_COMMON_HANDL r_bsn_cb;                       //11.read bsn
    TYTEST_COMMON_HANDL_NOBUF w_sn_cb;                  //12.write sn
    TYTEST_COMMON_HANDL r_sn_cb;                        //13.read sn
    TYTEST_COMMON_HANDL_NOBUF w_country_cb;             //14.write country
    TYTEST_COMMON_HANDL r_country_cb;                   //15.read country
    TYTEST_KEY_HANDL key_cb;                            //16.key test
    TYTEST_COMMON_HANDL_NOBUF led_cb;                   //17.led test   
    TYTEST_COMMON_ASYNC_HANDL slave_cap_cb;                   //18.slave cap test      
    TYTEST_COMMON_HANDL_NOBUF upgrade_slave_cb;         //19.slave upgrade      
    TYTEST_COMMON_HANDL_NOBUF upgrade_cfg_cb;           //20.upgrade cfg   
    TYTEST_COMMON_HANDL_NOBUF upgrade_start_cb;         //21.upgrade start    
    TYTEST_COMMON_HANDL_NOBUF upgrade_rev_packet_cb;    //22.upgrade receive     
    TYTEST_COMMON_HANDL_NOBUF upgrade_end_cb;           //23.upgrade end   
    TYTEST_REBOOT_HANDL reboot_cb;                      //24.reboot
    TYTEST_COMMON_HANDL_NOBUF speaker_cb;               //25.speaker
    TYTEST_COMMON_HANDL lte_test_cb;                    //26.lte test
    TYTEST_COMMON_HANDL r_ccid_cb;                      //27.read ccid
    TYTEST_COMMON_HANDL flag_dect_cb;                   //28.flag dectect
    TYTEST_CUSTOMIZE_HANDL customize_db;                //29.customize handle    
    TYTEST_KEY_ASYNC_HANDL key_async_cb;                //30.key test async rep
}TUYA_TEST_CBS_S;

//产测主控制结构体
typedef struct tuya_prodtest_handler{
    int prod_start_flag;
	TUYA_PRODTEST_MODE trans_mode;
	int tcp_port;
    TUYA_TEST_THREAD_STATE thd_run_state;
	TF_HANDLER frame_handle; 
    TUYA_TEST_CBS_S cbs;    
}TP_HANDLER;

//function
int tuya_prodtest_rep_key_event(int index, unsigned int key_num);
int tuya_prodtest_rep_zigbeeRf_event(int index, unsigned int receive_num);
void tuya_prodtest_set_tcp_port(int tcp_port);
void tuya_prodtest_set_start_flag(bool status);
int tuya_prodtest_init(TUYA_PRODTEST_MODE mode, TUYA_TEST_CBS_S *cbs);
int tuya_prodtest_deinit();


#ifdef __cplusplus
}
#endif

#endif



Library/gcc/sdk/include/tuya_testframe_handle.h
#ifndef _TUYA_TESTFRAME_HANDLE_H_
#define _TUYA_TESTFRAME_HANDLE_H_

#ifdef __cplusplus
extern "C"
{
#endif

#include "tuya_prod_test.h"

#define TYPT_FRAME_DATA_MAX_LEN 1048
#define falseStr "{\"ret\":false}"
#define trueStr  "{\"ret\":true}"

/*
    次固件测试支持的类型
    zigbee wifi ble
*/
typedef enum {
    WT_ZIGBEE = 1,
    WT_WIFI = 2,
    WT_BLE = 3
}WIRELESS_TYPE;


//灯状态
typedef enum {
    TYPT_LED_ALL_ON = 0,
    TYPE_LED_ALL_OFF,
    TYPT_LED_FLASH
}TYPT_LED_STATE;

//扬声器的开关
typedef enum {
    TYPT_SOUND_OFF = 0,    
    TYPT_SOUND_ON
}TYPT_SOUND_STATE;

//lte测试步骤
typedef enum{
    TYPT_LTE_ENTER_TEST = 1,
    TYPT_LTE_EXIT_TEST,
    TYPT_LTE_TEST_MODE1,    
    TYPT_LTE_TEST_MODE2,
    TYPT_LTE_TEST_MODE3,
}TYPT_LTE_TEST;

//板载flag检测
typedef enum{
    TYPT_FLAG_WIFIRF = 0,    
    TYPT_FLAG_BOARDTEST = 1,
}TYPT_FLAG_TYPE;

//次固件升级类型
typedef enum {
    UG_ZIGBEE_IMAGE = 1,    
    UG_ZWAVE_IMAGE = 2,
    UG_BLE_IMAGE = 3,
    UG_MAX,
}UPGRADE_TYPE;

//CALLBACK
typedef void (*TYTEST_F_COMMON_VOID_HANDL)(void);
typedef int (*TYTEST_F_COMMON_HANDL)(void);
typedef int (*TYTEST_F_WR_HANDL)(char *buf);
typedef int (*TYTEST_F_W_CFG_HANDL)(char *auzkey, char *uuid, char *pid, int prodtest, char *ap_ssid, char *ap_pwd);
typedef int (*TYTEST_F_R_CFG_HANDL)(char *auzkey, char *uuid, char *pid, int *prodtest, char *ap_ssid, char *ap_pwd);
typedef int (*TYTEST_F_WR_SLAVEMAC_HANDL)(int cpunum,char *mac);
typedef int (*TYTEST_F_KEY_HANDL)(unsigned int *response_key_num);
typedef void (*TYTEST_F_LED_HANDL)(TYPT_LED_STATE state);
typedef int (*TYTEST_F_ZIGBEE_RF_HANDL)(int index,int channel, int num);
typedef int (*TYTEST_F_WIFIBLE_RF_HANDL)(char *ssid, int *rssi);
typedef int (*TYTEST_F_UG_SLAVE_HANDL)(UPGRADE_TYPE image_type);
typedef int (*TYTEST_F_UG_CFG_HANDL)(unsigned long baud, unsigned long crc32, unsigned long size);
typedef int (*TYTEST_F_UG_REV_HANDL)(char *data, unsigned int data_len);
typedef void (*TYTEST_F_SPEAKER_HANDL)(TYPT_SOUND_STATE state);
typedef int (*TYTEST_F_LTE_HANDL)(TYPT_LTE_TEST state);
typedef int (*TYTEST_F_R_CCID_HANDL)(char *ccid);
typedef int (*TYTEST_F_R_FLAGTYPE_HANDL)(TYPT_FLAG_TYPE type);
typedef int (*TYTEST_F_CUSTOMIZE_HANDL)(char *data, char *out_buf);
typedef int (*TYTEST_F_KEY_ASYNC_HANDL)(int index);


typedef struct {    
    TYTEST_F_COMMON_HANDL enter_prodtest_frame_cb;          //1.enter prodtest    
    TYTEST_F_COMMON_HANDL exit_prodtest_frame_cb;           //1.exit prodtest
    TYTEST_F_WR_HANDL master_firm_frame_cb;                 //2.master_firm
    TYTEST_F_WR_HANDL slave_firm_frame_cb;                  //3.slave firm
    TYTEST_F_W_CFG_HANDL w_cfg_frame_cb;                    //4.write cfd info
    TYTEST_F_R_CFG_HANDL r_cfg_frame_cb;                    //5.read cfg info
    TYTEST_F_WR_HANDL w_master_mac_frame_cb;                //6.write master mac
    TYTEST_F_WR_HANDL r_master_mac_frame_cb;                //7.read master mac
    TYTEST_F_WR_SLAVEMAC_HANDL w_slave_mac_frame_cb;        //8.write slave mac
    TYTEST_F_WR_SLAVEMAC_HANDL r_slave_mac_frame_cb;        //9.read slave mac
    TYTEST_F_WR_HANDL w_bsn_frame_cb;                       //10.write bsn
    TYTEST_F_WR_HANDL r_bsn_frame_cb;                       //11.read bsn
    TYTEST_F_WR_HANDL w_sn_frame_cb;                        //12.write sn
    TYTEST_F_WR_HANDL r_sn_frame_cb;                        //13.read sn
    TYTEST_F_WR_HANDL w_country_frame_cb;                   //14.write country
    TYTEST_F_WR_HANDL r_country_frame_cb;                   //15.read country
    TYTEST_F_KEY_HANDL key_frame_cb;                        //16.key test
    TYTEST_F_LED_HANDL led_frame_cb;                        //17.led test 
    TYTEST_F_ZIGBEE_RF_HANDL zigbee_rf_frame_cb;            //18.zigbee wifi ble rf test  
    TYTEST_F_WIFIBLE_RF_HANDL wifi_rf_frame_cb;    
    TYTEST_F_WIFIBLE_RF_HANDL ble_rf_frame_cb;
    TYTEST_F_UG_SLAVE_HANDL upgrade_slave_frame_cb;         //19.slave upgrade      
    TYTEST_F_UG_CFG_HANDL upgrade_cfg_frame_cb;             //20.upgrade cfg   
    TYTEST_F_COMMON_HANDL upgrade_start_frame_cb;           //21.upgrade start    
    TYTEST_F_UG_REV_HANDL upgrade_rev_packet_frame_cb;      //22.upgrade receive     
    TYTEST_F_COMMON_HANDL upgrade_end_frame_cb;             //23.upgrade end   
    TYTEST_F_COMMON_VOID_HANDL reboot_frame_cb;             //24.reboot
    TYTEST_F_SPEAKER_HANDL speaker_frame_cb;                //25.speaker
    TYTEST_F_LTE_HANDL lte_test_frame_cb;                   //26.lte test
    TYTEST_F_R_CCID_HANDL r_ccid_frame_cb;                  //27.read ccid
    TYTEST_F_R_FLAGTYPE_HANDL flag_dect_frame_cb;           //28.flag dectect
    TYTEST_F_CUSTOMIZE_HANDL customize_frame_cb;            //29.customize handle    
    TYTEST_F_KEY_ASYNC_HANDL key_async_frame_cb;            //30.key test
}TYTEST_FRAME_CBS_S;

//产测帧处理主控制结构体
typedef struct tuya_test_frame_handler{
	int register_status;
    TYTEST_FRAME_CBS_S cbs;
}TYTEST_FRAME_HANDLER;

//function
int tuya_testframe_rep_key_event(int index, unsigned int key_num);
int tuya_testframe_rep_zigbeeRf_event(int index, unsigned int receive_num);
void tuya_testframe_set_tcp_port(int tcp_port);
int tuya_testframe_handle_init(TUYA_PRODTEST_MODE mode, TYTEST_FRAME_CBS_S *cbs);
int tuya_testframe_handle_deinit();


#ifdef __cplusplus
}
#endif

#endif



Library/msdk-4.4.7-mips-EL-3.10-u0.9.33-m32t-140827/demo/demo.c
#include <unistd.h>
#include <stdio.h>
#include <pthread.h>
#include "tuya_testframe_handle.h"

//线程创建接口
static int creat_thread_and_start(pthread_t *thrid, thread_process thr_process, void *args)
{
    int opRet = 0;
    pthread_attr_t attr;
    
    pthread_attr_init(&attr);
    pthread_attr_setdetachstate(&attr,PTHREAD_CREATE_DETACHED);
    opRet = pthread_create(thrid,&attr, thr_process, args);
    pthread_attr_destroy(&attr);
    if(opRet != 0)
    {
        TYPD_ERR("Create Thread Fails %d", opRet);
        return -1;
    }  
    return 0;
}

//线程销毁
static int cancel_thread(pthread_t thrid)
{
#ifdef __ANDROID__
    //NOTE:
    //安卓不支持 pthread_cancel
    //如果被其他线程退出，安卓版本该接口暂不支持
    pthread_exit((void*)thrid);
    return 0;
#else
    return pthread_cancel(thrid);
#endif
}

/*进入产测模式*/
static int tuya_prod_enter(void)
{
    printf("enter prodtest\n");
    return 0;
}

/*退出产测模式*/
static int tuya_prod_exit(void)
{
    printf("exit prodtest\n");
    return 0;
}

/*读取主固件版本信息*/
static int tuya_prod_r_master_firmware_info(char *out_buf)
{
    if(out_buf == NULL)
        return -1;
    sprintf(out_buf,"{\"ret\":true,\"firmName\":\"home_security_gw\",\"firmVer\":\"1.2.3\"}");
    return 0;
    
}

/*读取次固件版本信息*/
static int tuya_prod_r_slave_firmware_info(char *out_buf)
{
    if(out_buf == NULL)
        return -1;
    sprintf(out_buf,"{\"ret\":true,\"firmName\":\"z3_ncp_firmware\",\"firmVer\":\"1.0.6\"}");
    return 0;
}

/*写入配置信息*/
static int tuya_prod_w_auth_info(char *auzkey, char *uuid, char *pid, int prodtest, char *ap_ssid, char *ap_pwd)
{
    printf("auzkey:%s  uuid:%s prodtest:%d\n",auzkey,uuid,prodtest);
    if(pid)        
        printf("pid:%s\n",pid);
    if(ap_ssid)        
        printf("apssid:%s\n",ap_ssid);
    if(ap_pwd)        
        printf("appwd:%s\n",ap_pwd);

    return 0;
}

/*读取配置信息*/
static int tuya_prod_r_auth_info(char *auzkey, char *uuid, char *pid, int *prodtest, char *ap_ssid, char *ap_pwd)
{
    sprintf(auzkey,"U4Ri3GGd6VOR255oumyfgLPbJ8ITtH8W");    
    sprintf(uuid,"bc85a0edd3458fce");
    //sprintf(pid,"kjkjkjkjsasa");
    *prodtest = 0;    
    //sprintf(ap_ssid,"smartlife-0009");    
    //sprintf(ap_pwd,"buzhidao");
    return 0;        
}

/*写入主mac地址*/
static int tuya_prod_w_master_mac(char *mac)
{
    printf("master mac is:%s\n",mac);
    return 0;
}

/*读取主mac地址*/
static int tuya_prod_r_master_mac(char *mac)
{
    sprintf(mac,"080027a2b63c");
    return 0;    
}

/*写入次mac地址*/
static int tuya_prod_w_slave_mac(int cpunum,char *mac)
{
    printf("slave mac%d is:%s\n",cpunum,mac);
    return 0;
}

/*读取次mac地址*/
static int tuya_prod_r_slave_mac(int cpunum,char *mac)
{
    printf("read cpunum %d\n",cpunum);
    sprintf(mac,"080027a2b64d");
    return 0;    
}

/*写入bsn*/
static int tuya_prod_w_bsn(char *bsn)
{
    printf("bsn is:%s\n",bsn);
    return 0;
}

/*读取bsn*/
static int tuya_prod_r_bsn(char *bsn)
{
    sprintf(bsn,"bsn1234567");
    return 0;    
}

/*写入sn*/
static int tuya_prod_w_sn(char *sn)
{
    printf("sn is:%s\n",sn);
    return 0;
}

/*读取sn*/
static int tuya_prod_r_sn(char *sn)
{
    sprintf(sn,"sn12345678");
    return 0;    
}

/*写入国家码*/
static int tuya_prod_w_country_code(char *country_code)
{
    printf("country_code is:%s\n",country_code);
    return 0;
}

/*读取国家码*/
static int tuya_prod_r_country_code(char *country_code)
{
    sprintf(country_code,"CN");
    return 0;    
}

/*
    led测试：亮 灭 闪烁
*/
void tuya_prod_ledtest(TYPT_LED_STATE state)
{
    printf("led blink\n");
}

/*
    按键测试
    函数说明：返回值为按键测试成功的个数，可阻塞，比如板子上有两个按键，
              如果测试两个按键都没有问题，*response_key_num = 2即可
*/
static int tuya_prod_keytest(unsigned int *response_key_num)
{
    *response_key_num = 2;
    return 0;
}

static int zb_index[1024] = {0};
static void* tuya_prod_zbrf_thread_process(void *pArg)
{
    int *index = (int*)pArg;
    sleep(5);

    TYPD_DEBUG("*****************************ZBRF INDEX:%d\n",*index);
    tuya_testframe_rep_zigbeeRf_event(*index,100);
    
    return (void*)0;
}


/*
    zigbee rf测试
    函数说明：异步上报；入参为index（异步上报需要用到，和上报用到的index一一对应，是同一个值），信道，发包的个数，出参为收到的报文个数
*/
static int tuya_prod_zigbee_rf_test(int index, int channel, int num)
{
    printf("index:%d ,zigbee rf test:channel %d,num:%d",index,channel, num);    
    //thread 
	pthread_t pt_thr;
	int ret = 0;

    zb_index[index] = index;
	
	ret = creat_thread_and_start(&pt_thr,tuya_prod_zbrf_thread_process,(void*)(&zb_index[index]));
    if(0 != ret)
    {
        TYPD_ERR("start tuya_prod_key_thread_process fails %d", ret);
        return -1;
    }
    return 0;
}

/*
    wifi rf测试
    函数说明：可阻塞；入参为ssid，出参为该ssid的信号强度
*/
static int tuya_prod_wifi_rf_test(char *ssid, int *rssi)
{
    printf("wifi rf test:%s\n",ssid);
    *rssi = -30;
    return 0;
}

/*
    ble rf测试
    函数说明：可阻塞；入参为ssid，出参为该ssid的信号强度
*/
static int tuya_prod_ble_rf_test(char *ssid, int *rssi)
{
    printf("ble rf test:%s\n",ssid);
    *rssi = -20;
    return 0;
}

/*
    扬声器测试：开，关
*/
static void tuya_prod_speaker_test(TYPT_SOUND_STATE state)
{
    printf("speaker test!\n");
}

/*
    lte测试
    函数说明：入参为lte测试模式的进入，退出，及三个频段测试模式，用户要根据当前传进来的模式做具体的方案实施
*/
static int tuya_prod_lte_test(TYPT_LTE_TEST state)
{
    printf("lte test:%d\n",state);
    return 0;
}

/*
    板载flag测试：目前只有rokidk18平台用到
*/
static int tuya_prod_r_flagtype(TYPT_FLAG_TYPE type)
{
    printf("%d flagtype read",type);
    return 0;
}

/*重启系统*/
static void tuya_prod_reboot(void)
{
    printf("reboot now!\n");
}

/*
    自定义产测协议接口
    函数说明：入参为产测帧的data字段，用户需要自行解析；
              返回给上位机的产测帧的data部分放到out_buf中
*/
static int tuya_prod_customize(char *data, char *out_buf)
{    
    if(data)    
        printf("data is %s\n",data);
    sprintf(out_buf,"customize handle");
    return 0;
}

/*读取ccid*/
static int tuya_prod_r_ccid(char *ccid)
{   
    printf("read ccid\n");    
    sprintf(ccid,"123456789012345678");
    return 0;
}

/*选择进行次固件产测升级*/
static int tuya_prod_upgrade_slaveimg_cfg(UPGRADE_TYPE image_type)
{
    printf("upgrade slave:%d\n",image_type);
    return 0;
}

/*产测升级配置：入参：串口波特率，固件crc32校验，固件大小*/
static int tuya_prod_upgrade_cfg(unsigned long baud, unsigned long crc32, unsigned long size)
{
    printf("upgrade cfg:%lul %lul %lul\n",baud, crc32, size);
    return 0;
}

/*产测升级开始*/
static int tuya_prod_upgrade_start(void)
{
    printf("upgrade start\n");
    return 0;
}

/*产测升级接受数据*/
static int tuya_prod_upgrade_rev(char *data, unsigned int data_len)
{    
    printf("upgrade receive data len:%d\n",data_len);
    return 0;
}

/*产测升级结束*/
static int tuya_prod_upgrade_end(void)
{    
    printf("upgrade end\n");
    return 0;
}

static int key_index[1024] = {0};
static void* tuya_prod_key_thread_process(void *pArg)
{
    int *index = (int*)pArg;
    sleep(5);
    
    TYPD_DEBUG("*****************************KEY INDEX:%d\n",*index);
    tuya_testframe_rep_key_event(*index,3);    

    return (void*)0;
}

/*
    按键测试异步
    函数说明：入参为index（异步上报需要用到，和上报用到的index一一对应，是同一个值）
*/
static int tuya_prod_key_async_test(int index)
{
    //thread 
	pthread_t pt_thr;
	int ret = 0;
    static int sindex[1024] = {0};
    
    key_index[index] = index;
	
	ret = creat_thread_and_start(&pt_thr,tuya_prod_key_thread_process,(void*)(&key_index[index]));
    if(0 != ret)
    {
        TYPD_ERR("start tuya_prod_key_thread_process fails %d", ret);
        return -1;
    }
    return 0;
}

/*
    老化测试检测，返回0 说明进行过老化测试，-1说明没有
*/
static int tuya_prod_ageing_test()
{
    return 0;
}

int main()
{
    TYTEST_FRAME_CBS_S prodframe_cbs = {
        .enter_prodtest_frame_cb = tuya_prod_enter,                            //1.enter prodtest    
        .exit_prodtest_frame_cb = tuya_prod_exit,                             //1.exit prodtest
        .master_firm_frame_cb = tuya_prod_r_master_firmware_info,           //2.master_firm
        .slave_firm_frame_cb = tuya_prod_r_slave_firmware_info,            //3.slave firm
        .w_cfg_frame_cb = tuya_prod_w_auth_info,                      //4.write cfd info
        .r_cfg_frame_cb = tuya_prod_r_auth_info,                      //5.read cfg info
        .w_master_mac_frame_cb = tuya_prod_w_master_mac,                     //6.write master mac
        .r_master_mac_frame_cb = tuya_prod_r_master_mac,                     //7.read master mac
        .w_slave_mac_frame_cb = tuya_prod_w_slave_mac,                      //8.write slave mac
        .r_slave_mac_frame_cb = tuya_prod_r_slave_mac,                      //9.read slave mac
        .w_bsn_frame_cb = tuya_prod_w_bsn,                            //10.write bsn
        .r_bsn_frame_cb = tuya_prod_r_bsn,                            //11.read bsn
        .w_sn_frame_cb = tuya_prod_w_sn,                             //12.write sn
        .r_sn_frame_cb = tuya_prod_r_sn,                             //13.read sn
        .w_country_frame_cb = tuya_prod_w_country_code,                   //14.write country
        .r_country_frame_cb = tuya_prod_r_country_code,                   //15.read country
        .key_frame_cb = tuya_prod_keytest,                          //16.key test
        .led_frame_cb = tuya_prod_ledtest,                          //17.led test 
        .zigbee_rf_frame_cb = tuya_prod_zigbee_rf_test,                   //18.zigbee wifi ble rf test  
        .wifi_rf_frame_cb = tuya_prod_wifi_rf_test,
        .ble_rf_frame_cb = tuya_prod_ble_rf_test,
        .upgrade_slave_frame_cb = tuya_prod_upgrade_slaveimg_cfg,             //19.slave upgrade      
        .upgrade_cfg_frame_cb = tuya_prod_upgrade_cfg,                      //20.upgrade cfg   
        .upgrade_start_frame_cb = tuya_prod_upgrade_start,                    //21.upgrade start    
        .upgrade_rev_packet_frame_cb = tuya_prod_upgrade_rev,                      //22.upgrade receive     
        .upgrade_end_frame_cb = tuya_prod_upgrade_end,                      //23.upgrade end   
        .reboot_frame_cb = tuya_prod_reboot,                           //24.reboot
        .speaker_frame_cb = tuya_prod_speaker_test,                     //25.speaker
        .lte_test_frame_cb = tuya_prod_lte_test,                         //26.lte test
        .r_ccid_frame_cb = tuya_prod_r_ccid,                            //27.read ccid
        .flag_dect_frame_cb = tuya_prod_r_flagtype,                       //28.flag dectect
        .customize_frame_cb = tuya_prod_customize,                        //29.customize handle
        .key_async_frame_cb = tuya_prod_key_async_test,
        .ageing_test_frame_cb = tuya_prod_ageing_test,         //31.ageing test
    };

	int ret;

	//如果不调用此接口设置端口号 端口号默认是12130
	//tuya_testframe_set_tcp_port(12131);
	
	ret = tuya_testframe_handle_init(TPM_TCP,&prodframe_cbs);
	if(ret < 0) {
	    printf("tuya_testframe_handle_init failed\n");
	    return -1;
	}
    	
	while(1) {
		sleep(1000);
	}
	return 0;
}

